src/Compilers/CSharp/Portable/Binder/Binder_Patterns.cs

```diff
@@ -431,6 +437,30 @@ private static BoundPattern BindDiscardPattern(DiscardPatternSyntax node, TypeSy
             }
         }
 
+        private bool HasBlockedINumberConversion(Conversion patternConversion, TypeSymbol inputType)
+        {
+            // We want to block constant and relation patterns that have an input type constrained to or inherited from INumberBase<T>, if we don't
+            // know how to represent the constant being matched against in the input type. For example, `1.0 is 1` will work when written inline, but
+            // will fail if the input type is `INumberBase<T>`. We block this now so that we can make make it work as expected in the future without
+            // being a breaking change.
+
+            if (patternConversion.IsIdentity || patternConversion.IsConstantExpression || patternConversion.IsNumeric)
+            {
+                return false;
+            }
+
+            var iNumberBase = Compilation.GetWellKnownType(WellKnownType.System_Numerics_INumberBase_T);
+
+            if (iNumberBase.IsErrorType())
+            {
+                return false;
+            }
+
+            var interfaces = inputType is TypeParameterSymbol typeParam ? typeParam.EffectiveInterfacesNoUseSiteDiagnostics : inputType.AllInterfacesNoUseSiteDiagnostics;
+            return interfaces.Any(static (i, iNumberBase) => i.OriginalDefinition.Equals(iNumberBase, TypeCompareKind.AllIgnoreOptions), iNumberBase)
+                   || inputType.OriginalDefinition.Equals(iNumberBase, TypeCompareKind.AllIgnoreOptions);
```

Comment:
>inputType.OriginalDefinition.Equals(iNumberBase, TypeCompareKind.AllIgnoreOptions) [](http://example.com/codeflow?start=22&length=82)

What happens if this check succeeds, but the type is not interface?

Reply:
> What happens if this check succeeds, but the type is not interface?

The spec doesn't say anything about whether `INumberBase<TSelf>` needs to be an interface, so I would expect this check to succeed.

Reply:
> so I would expect this check to succeed.

Please add a test for a scenario like that.